### PR TITLE
Implement and use `JaxBaseModuleClass`

### PR DESCRIPTION
**Pull request by adamgayoso**

### CODE REVIEW

File: scvi/model/_jaxscvi.py [hunk 1]
```diff
@@ -132,6 +132,7 @@ def module(self):
     def train(
         self,
         max_epochs: Optional[int] = None,
+        check_val_every_n_epoch: Optional[int] = None,
```
Comment (jjhong922) [inline scvi/model/_jaxscvi.py:135]: should we always append new kwargs to the end of the signature unless it really doesnt make sense? Just to avoid situations when they are used as args.

File: scvi/model/_jaxscvi.py [hunk 2]
```diff
@@ -191,32 +198,33 @@ def key(i: int):
...
-            apply_fn=module.apply,
+            apply_fn=train_module.apply,
             params=params,
             tx=optax.adamw(lr, eps=0.01, weight_decay=1e-6),
             batch_stats=batch_stats,
         )
 
         @jax.jit
-        def train_step(state, array_dict, rngs, kl_weight=1) -> jnp.ndarray:
+        def train_step(state, array_dict, rngs, kwargs):
```
Comment (jjhong922) [inline scvi/model/_jaxscvi.py:213]: is **kwargs not allowed in jit?

File: scvi/module/base/_base_module.py [hunk 3]
```diff
@@ -36,14 +38,35 @@ class LossRecorder:
 
     def __init__(
         self,
-        loss: Union[Dict[str, torch.Tensor], torch.Tensor],
+        loss: Union[
+            Dict[str, Union[torch.Tensor, jnp.ndarray]],
```
Comment (jjhong922) [inline scvi/module/base/_base_module.py:42]: I should really address https://github.com/scverse/scvi-tools/issues/1131

File: scvi/module/base/_base_module.py [hunk 4]
```diff
@@ -69,19 +92,19 @@ def _get_dict_sum(dictionary):
         return total
 
     @property
-    def loss(self) -> torch.Tensor:
```
Comment (jjhong922) [inline scvi/module/base/_base_module.py:72]: Might be overkill, but at the loss recorder level it would be nice if the values were just stored agnostic to the storage type (tensor vs jnp array).

File: scvi/module/base/_base_module.py [hunk 5]
```diff
@@ -343,3 +366,133 @@ def create_predictive(
     def forward(self, *args, **kwargs):
         """Passthrough to Pyro model."""
         return self.model(*args, **kwargs)
+
+
+class JaxBaseModuleClass(linen.Module):
+    """Abstract class for Jax-based scvi-tools modules."""
+
+    def __init__(*args, **kwargs):
+        super().__init__(**kwargs)
```
Comment (jjhong922) [inline scvi/module/base/_base_module.py:375]: why doesnt *args get passed here?

File: scvi/module/base/_base_module.py [hunk 6]
```diff
@@ -343,3 +366,133 @@ def create_predictive(
...
+        In the case of variational inference, this function will perform steps related to
+        computing variational distribution parameters. In a VAE, this will involve running
+        data through encoder networks.
+
+        This function should return a dictionary with str keys and :class:`~jnp.ndarray` values.
+        """
+        pass
+
+    @abstractmethod
+    def generative(self, *args, **kwargs) -> dict:
```
Comment (jjhong922) [inline scvi/module/base/_base_module.py:477]: For whole file, type the Dict return values

**Review by jjhong922 — COMMENTED** [review]
Overall looks really good. JaxBaseModule is nicely consistent with BaseModule. It is a shame that beyond looking similar we can't tie together the pytorch and jax components (probably mostly bc of pytorch lightning). It would be cool if everything was agnostic to the backend with the exception of the very lowest level of calling flax modules.

File: scvi/model/_jaxscvi.py (hunk 1 above)
Comment (adamgayoso) [reply scvi/model/_jaxscvi.py:135]: yes, but this whole API is experimental. Right now for scvi et al. this is part of **kwargs to the trainer. I think this is fine for now.

File: scvi/module/base/_base_module.py (hunk 3 above)
Comment (adamgayoso) [reply scvi/module/base/_base_module.py:42]: yes later :)

File: scvi/module/base/_base_module.py (hunk 5 above)
Comment (adamgayoso) [reply scvi/module/base/_base_module.py:375]: because this is some randomly generated vscode fill in

File: scvi/module/base/_base_module.py [hunk 7]
```diff
@@ -343,3 +358,155 @@ def create_predictive(
...
+
+        This function uses the outputs of the inference and generative functions to compute
+        a loss. This many optionally include other penalty terms, which should be computed here.
+
+        This function should return an object of type :class:`~scvi.module.base.LossRecorder`.
+        """
+        pass
+
+
+def _generic_forward(
```
Comment (jjhong922) [inline scvi/module/base/_base_module.py:481]: where else is this used?

**Comment by adamgayoso** [comment]
@jjhong922 the error here is due to amortized lda save load error precision

File: scvi/module/base/_base_module.py (hunk 7 above)
Comment (adamgayoso) [reply scvi/module/base/_base_module.py:481]: ? Only inside this file, both by jax call and pytorch forward

Comment (jjhong922) [reply scvi/module/base/_base_module.py:481]: Oh ok missed the pytorch one. Just wanted to warn against refactoring too early.

**Comment by jjhong922** [comment]
@adamgayoso just rerunning it to see if its truly a flake. But I suppose we can manually increase the margin on the test too
